### ISSUE TITLE
cmplog colorization type_replace often not random

### ISSUE THREAD

**Issue by guyf2010** [AFLplusplus/AFLplusplus#1546]
The type_replace function used to in cmplog's colorization step often does not mutate randomly depending on the data in the current byte.

The switch statement at the following location contains many data values, but a large number of destinations in the switch do not apply any sort of randomness to their replacements. https://github.com/AFLplusplus/AFLplusplus/blob/7e0171006a32bf6b510e08df76ca6d8555272140/src/afl-fuzz-redqueen.c#L182-L254

In particular, any byte containing a value in the range from 128 to 254 will simply be XORed with 0x7F. I assume the intention was to XOR with a random value from 0 to 0x7F.

Additionally, the original RedQueen paper uses an EXT4 filesystem as a motivating example for colorization, where more than 60,000 instances of the value 0 were present, and colorization can greatly reduce the number of candidate replacements. However, the current cmplog implementation can only replace a value of 0 with 1, making it far less effective in this use case. 

In my testing the input-to-state stage for a single seed can run for extremely long periods of time (Several minutes to hours in some cases). Discovering seeds prone to this seems particularly common from the clone mutation with a constant value. https://github.com/AFLplusplus/AFLplusplus/blob/7e0171006a32bf6b510e08df76ca6d8555272140/src/afl-fuzz-one.c#L2476-L2517

I'm unsure whether this is a bug or an intentional trade-off made for certain test cases. If it is an intentional decision, perhaps some comments should be added to those switch cases explaining why colorization does not randomize those byte values. The type_replace does seem very focused on ASCII, maybe a config option can be added for cmplog to not use these types when a non-ASCII based program is being fuzzed.

Thanks,
Guy

**Comment by vanhauser-thc** [comment]
Everything in that function is intentional, also the xor with a fixed value in specific cases.

I am on my phone so can’t write a thesis long rational now, but as with most in fuzzing it is guesswork and experience gathering with targets (fuzzbench).

if you want then make a pr that you think would be better and I do a fuzzbench comparison run.

Theories never help to know what is better in fuzzing, only real world testing :)

**Comment by vanhauser-thc** [comment]
> if you want then make a pr that you think would be better and I do a fuzzbench comparison run.

@guyf2010 any thinking about this? would you want to try?

**Comment by guyf2010** [comment]
> > if you want then make a pr that you think would be better and I do a fuzzbench comparison run.
> 
> @guyf2010 any thinking about this? would you want to try?

From what I've seen so far, the issue is dependent on the tested program. The programs I have tested with that encounter the issue, I cannot release. I recognise this makes it a difficult problem to demonstrate. Additionally, if Fuzzbench doesn't have any  binaries prone to these issues, no benefits can be shown from the benchmark.

I spoke to a colleague who said he had observed input-to-state being stuck for multiple hours on a single input for the 'who' binary from the LAVA set. I had a look at running that binary to post some results here, but the build dependencies are very outdated and not compatible with my current system, so I haven't been able to run it.

For my local system, I have reverted to just doing `rand_below(afl, 256)` on each byte, which resolved my issues.

I'm not sure what the goals of the project are here, the implementation seems very ASCII focused, rather than format generic. Only having one candidate replacement value for the majority of byte values doesn't seem to align my understanding of the purpose of colorization from reading the RedQueen paper.

While I'm open to writing a PR, I still don't understand why it's written the way it is presently. So perhaps I'm not well qualified to make the changes. If I get a chance in the future I might be able to look at the Fuzzbench binaries and see how these optimisations apply to those sorts of programs. Until then, perhaps all I can do is report in rare cases, AFL++ spends hours in the input-to-state stage for a single input.

Thanks for your time,
Guy
